apis/r/R/SOMANDArrayBase.R

```diff
@@ -84,7 +84,20 @@ SOMANDArrayBase <- R6::R6Class(
     set_data_type = function(type) {
       spdl::debug("[SOMANDArrayBase::set_data_type] caching type {}", type$ToString())
       private$.type <- type
+    },
+
+    #' @description Returns TRUE if the array has the upgraded resizeable shape
+    #' feature from TileDB-SOMA 1.14: the array was created with this support,
+    #' or it has had ``upgrade_domain`` applied to it.
+    #' (lifecycle: maturing)
+    #' @return Logical
+    has_upgraded_shape = function() {
+      has_current_domain(
+        self$uri,
+        config=as.character(tiledb::config(self$tiledbsoma_ctx$context()))
+      )
     }
+
```

Comment:
Is there a reason we're using a different name for NDArrays vs data frames, despite both running the same code? Can we agree on a single name for all array-like objects and push this down to a lower-level class (eg. `SOMAArrayBase` or `TileDBArray`) to avoid code duplication?

Reply:
Yes, please see the committee-agreed namings on #2407.

To summarize:

* `SOMASparseNDArray` and `SOMADenseNDArray` must have dims of type `int64` -- only
* `SOMADataFrame` must have a dim or attr named `soma_joinid` and it must be of type `int64`
  * People can have `soma_joinid` as dim by itself (this is the default behavior and by far the most common)
  * People can have `soma_joinid` as a dim along with one or more other dims -- and those can have arbitrary type
  * People can have one or more dims of arbitrary type while `soma_joinid is an attr
* This is why throughout the existing code we have `shape` in terms of array of `int64`, and `domain` is specified for `SOMADataFrame`. Also note that in #2407 we have `shape` and `maxshape` accessors for ND arrays but `domain` and `maxdomain` for dataframes. Likewise `upgrade_shape` vs `upgrade_domain`.

Given that the committee decision was hard-won and hammered out over many meetings and much discussion, I'm adhering to what was agreed on in this implementation.

Reply:
> * `SOMADataFrame` must have a dim or attr named `soma_joinid` and it must be of type `int64`
>   * People can have `soma_joinid` as dim by itself (this is the default behavior and by far the most common)
>   * People can have `soma_joinid` as a dim along with one or more other dims -- and those can have arbitrary type
>   * People can have one or more dims of arbitrary type while `soma_joinid` is an attr

BTW this is what `index_column_names` is all about

Reply:
> Yes, please see the committee-agreed namings

Good enough for me. I was not aware that these namings had been decided upon, nor was I apart of any committee making these decisions, but if that's already been hammered out then 👍 

> Re "despite running the same code" -- it's actually not. Given C++'s strong typing and use of classes [...]

Yeah, but from the R API side, that's not important. All I see is these two methods are
1. calling the same function (`has_current_domain()` with the same arguments (`self$uri` and `config = ...`)
2. have different names

What the function is doing (passing data to Rcpp, which then passes on to libtiledbsoma, which then dispatches based on the C++ class) is not important; all I see is code duplication (again, I was not aware of any committee to decide on these names). However, if a committee decided that these differing names are important, then 👍